### PR TITLE
Add ability to store multiple cells per HDF5

### LINKED ISSUE #61: Store multiple cells in the same HDF5 file
Some of our datasets involve thousands of cells, and creating thousands of files can be a problem

### PR DESCRIPTION
**Pull request by WardLT**
Fixes #61

- [x] Add warning if metadata differs between cells added to a H5
- [x] Add an option that reads any data from an h5, even if a user doesn't know the prefixes

### CODE REVIEW

File: batdata/data.py [hunk 1]
```diff
@@ -213,12 +212,19 @@ def from_batdata_hdf(cls, path_or_buf: Union[str, Path, HDFStore], subsets: Opti
...
-                raise ValueError(f'Unknown subset: {key}')
+        for subset in subsets:
+            # Throw error if user provides an unknown subset name
+            if subset not in _subsets:
+                raise ValueError(f'Unknown subset: {subset}')
+
+            # Prepend the prefix
+            if prefix is not None:
+                key = f'{prefix}-{subset}'
+            else:
```
Comment (victorventuri) [inline batdata/data.py:228]: If we want to get all the cells stored in the hdf5 file, regardless of 'prefix', does this work? Or would it raise an error? 
Also, doesn't this require one to know all the prefixes that have been stored?

File: docs/dataset.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# The `BatteryDataset` Object
```
Comment (victorventuri) [inline docs/dataset.md:1]: I just want to say I think this is a fantastic idea! A code-free full description of the main class we use will make it much easier for us to use and alter it as needs arise, as well as for other users to smoothly get started. Nicely done!

File: batdata/data.py [hunk 3]
```diff
@@ -234,6 +240,43 @@ def from_batdata_hdf(cls, path_or_buf: Union[str, Path, HDFStore], subsets: Opti
...
+            names = set()
+            for key in f.keys():
+                if '-' not in key:  # From the "default" group
+                    names.add(None)
+                name, _ = key.rsplit("-", 1)
+                names.add(name)
+            return metadata, names
+
+    @staticmethod
+    def get_metadata_from_hdf5(path: Union[str, Path]) -> BatteryMetadata:
```
Comment (victorventuri) [inline batdata/data.py:294]: I'm a little confused... Is this a different function from what it used to be?

**Review by victorventuri — COMMENTED** [review]
The main question that I have is: do these changes require one to know the prefixes that have been saved in order to read the data? 

The reason I ask is: in the "released" version, it is relatively straightforward; when I read a battdata.h5, I expect to have at least a raw_data and a metadata fields, and, in most occasions, a cycle_stats one as well. With these newer changes, it is not clear to me how I would get any data without knowing the prefix. In my understanding, the inspect_batdat can inform us which prefixes are stored, which definitely alleviates this.

File: batdata/data.py (hunk 1 above)
Comment (WardLT) [reply batdata/data.py:228]: Yea, it does require knowing the prefixes if they are used when saving the file. 

Maybe we should return an error if the prefix is not included in those found in the file?

File: batdata/data.py (hunk 3 above)
Comment (WardLT) [reply batdata/data.py:294]: Do you mean different than the old method for reading the metadata from an HDF file?

File: docs/dataset.md (hunk 2 above)
Comment (WardLT) [reply docs/dataset.md:1]: Yep, we totally need this for a bunch of features. Will gradually piece it together

File: batdata/data.py (hunk 1 above)
Comment (WardLT) [reply batdata/data.py:228]: I'm going to add that warning message

Comment (WardLT) [reply batdata/data.py:228]: Done

File: batdata/data.py (hunk 3 above)
Comment (victorventuri) [reply batdata/data.py:305]: Yeah, comparing the additions and deletions on Github it appears they are the same function? Maybe just moved to a different place in the file?

Comment (WardLT) [reply batdata/data.py:305]: Correct. It just moved

**Comment by WardLT** [comment]
Good point a out the potential annoyance of having to lookup prefixes before reading. How about I add an 'auto' option to the read which will grab data belonging to the first prefix then warn if there are more than one available?

**Comment by victorventuri** [comment]
I think the 'auto' idea is good, my question there is: how is the "first" prefix identified? I mean, there's no inherent order to them, right? 

I also think there should be a 'all' option, which basically loads all prefixes available. If you think it's wise, the 'all' can also output a warning for memory constraints.

**Comment by WardLT** [comment]
@victorventuri , I opted to add a "load by index" option to supply the need for "any." `prefix=0` will load the first prefix, alphabetically. 

Does this give you the features you'd need?

**Pull request merged**